### PR TITLE
WIP: Feature/async calls

### PR DESCRIPTION
**Pull request by DeviantBadge**

### CODE REVIEW

File: src/main/kotlin/io/github/rybalkinsd/kohttp/dsl/async/HttpPutAsyncDsl.kt [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+ *  }
+ * </pre>
+ *
+ * @see Response
+ * @see HttpPutContext
+ *
+ * @since 0.10.0
+ * @author evgeny
+ */
+fun httpPutAsync(
```
Comment (rybalkinsd) [inline src/main/kotlin/io/github/rybalkinsd/kohttp/dsl/async/HttpPutAsyncDsl.kt]: any reason to break lines in this way?
what is the default *kotlin* way to make line breaks here?

File: src/main/kotlin/io/github/rybalkinsd/kohttp/ext/CallExt.kt [hunk 2]
```diff
@@ -12,15 +12,15 @@ import kotlin.coroutines.suspendCoroutine
 internal fun Call.Factory.call(request: Request): Response = newCall(request).execute()
```
Comment (rybalkinsd) [inline src/main/kotlin/io/github/rybalkinsd/kohttp/ext/CallExt.kt]: no reason to change this file

File: src/main/kotlin/io/github/rybalkinsd/kohttp/ext/HttpContextExt.kt [hunk 3]
```diff
@@ -8,12 +8,15 @@ import java.net.URL
  * @author sergey
  */
 fun HttpContext.url(url: URL) {
-    if (url.protocol != "http" && url.protocol != "https") throw IllegalArgumentException("unexpected scheme: $scheme")
     scheme = url.protocol
+    if (scheme != "http" && scheme != "https")
```
Comment (rybalkinsd) [inline src/main/kotlin/io/github/rybalkinsd/kohttp/ext/HttpContextExt.kt]: what is the reason to change line breaks in this file?

File: src/main/kotlin/io/github/rybalkinsd/kohttp/ext/StringExt.kt [hunk 4]
```diff
@@ -61,8 +63,10 @@ fun String.httpGet(client: Call.Factory = defaultHttpClient): Response =
  * @since 0.1.0
  * @author sergey
  */
-fun String.asyncHttpGet(client: Call.Factory = defaultHttpClient): Deferred<Response> =
-    GlobalScope.async(context = Unconfined) {
-        client.suspendCall(Request.Builder().url(this@asyncHttpGet).build())
-    }
+fun String.httpGetAsync(
+        client: Call.Factory = defaultHttpClient
```
Comment (rybalkinsd) [inline src/main/kotlin/io/github/rybalkinsd/kohttp/ext/StringExt.kt]: what is the reason of reformatting code here?

File: src/main/kotlin/io/github/rybalkinsd/kohttp/pool/CoroutineExecutorService.kt [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+import kotlinx.coroutines.*
+import org.jetbrains.spek.meta.Experimental
+import java.util.concurrent.Executor
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.ForkJoinPool
+import kotlin.coroutines.CoroutineContext
+
+/**
+ * @Author: evgeny.vorobyev
```
Comment (rybalkinsd) [inline src/main/kotlin/io/github/rybalkinsd/kohttp/pool/CoroutineExecutorService.kt]: pls make a unique name everywhere

File: src/main/kotlin/io/github/rybalkinsd/kohttp/pool/CoroutineExecutorService.kt [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+ * @Since: 0.10.0
+ */
+@Experimental
+class CoroutineExecutorService(
```
Comment (rybalkinsd) [inline src/main/kotlin/io/github/rybalkinsd/kohttp/pool/CoroutineExecutorService.kt]: it is better to move experimental changes to the separate branch

File: src/main/kotlin/io/github/rybalkinsd/kohttp/util/json.kt [hunk 7]
```diff
@@ -41,7 +41,7 @@ class Json {
     }
 
     override fun toString(): String =
-        elements.joinToString(separator = ",", prefix = "{", postfix = "}") {
-            (k,v) -> """"$k":$v"""
```
Comment (rybalkinsd) [inline src/main/kotlin/io/github/rybalkinsd/kohttp/util/json.kt]: what is the reason to change this fragment?

File: src/test/kotlin/io/github/rybalkinsd/kohttp/Utils.kt [hunk 8]
```diff
@@ -11,3 +16,22 @@ fun assertResponses(expected: Map<String, String>, actual: JsonNode) {
...
+fun getSuccessfulResponsesAmount(
+        results: List<Deferred<Response>>
+): Int = runBlocking {
+    results.map { getCode(it) }
+            .filter { it == 200 }
+            .size
+}
+
+suspend fun getCode(
+        result: Deferred<Response>
```
Comment (rybalkinsd) [inline src/test/kotlin/io/github/rybalkinsd/kohttp/Utils.kt]: what is the reason for such line breaks?

File: src/test/kotlin/io/github/rybalkinsd/kohttp/client/ClientBuilderTest.kt [hunk 9]
```diff
@@ -1,15 +1,6 @@
 package io.github.rybalkinsd.kohttp.client
```
Comment (rybalkinsd) [inline src/test/kotlin/io/github/rybalkinsd/kohttp/client/ClientBuilderTest.kt]: no need for this change

File: src/test/kotlin/io/github/rybalkinsd/kohttp/ext/StringAsyncExtTest.kt [hunk 10]
```diff
@@ -8,14 +8,14 @@ import kotlin.test.assertEquals
 /**
  * @author sergey
  */
-class AsyncStringExtKtTest {
+class StringAsyncExtTest {
```
Comment (rybalkinsd) [inline src/test/kotlin/io/github/rybalkinsd/kohttp/ext/StringAsyncExtTest.kt]: What is the reason for rename?

File: src/test/kotlin/io/github/rybalkinsd/kohttp/performance/KhttpAsyncHttpGetPerformanceTest.kt [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+import io.github.rybalkinsd.kohttp.pool.CoroutineExecutorService
+import khttp.responses.Response
+import kotlinx.coroutines.*
+import org.junit.Test
+import java.util.concurrent.ForkJoinPool
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicInteger
+import kotlin.system.measureTimeMillis
+
+public class KhttpAsyncHttpGetPerformanceTest {
```
Comment (rybalkinsd) [inline src/test/kotlin/io/github/rybalkinsd/kohttp/performance/KhttpAsyncHttpGetPerformanceTest.kt]: this test is not related to the release, also better to move to the experimental branch

File: build.gradle.kts [hunk 12]
```diff
@@ -11,22 +11,28 @@ plugins {
 }
 
 group = "io.github.rybalkinsd"
-version = "0.8.0"
+version = "0.10.0-SNAPSHOT"
 
 repositories {
     mavenCentral()
+    maven {
+        url = uri("http://repo.spring.io/libs-release/")
```
[the remark this answers is not in the document]
Comment (DeviantBadge) [reply build.gradle.kts]: it need to be added to experimental branch
I added it to add 'khttp' dependency

File: build.gradle.kts [hunk 13]
```diff
@@ -11,22 +11,28 @@ plugins {
...
+    }
 }
 
 dependencies {
     implementation(kotlin("stdlib-jdk8"))
     implementation(kotlin("reflect"))
-    implementation("org.jetbrains.kotlinx", "kotlinx-coroutines-core", "1.0.1")
+    implementation("org.jetbrains.kotlinx", "kotlinx-coroutines-core", "1.2.1")
+    // https://mvnrepository.com/artifact/khttp/khttp
+    implementation("khttp:khttp:1.0.0")
```
[the remark this answers is not in the document]
Comment (DeviantBadge) [reply build.gradle.kts]: Khttp was added to compare kohttp async performance with khttp regarding to #61 issue

File: src/main/kotlin/io/github/rybalkinsd/kohttp/dsl/async/HttpPutAsyncDsl.kt (hunk 1 above)
Comment (DeviantBadge) [reply src/main/kotlin/io/github/rybalkinsd/kohttp/dsl/async/HttpPutAsyncDsl.kt]: Here what kotlin docs says:
https://kotlinlang.org/docs/reference/coding-conventions.html#function-formatting
```
fun longMethodName(
    argument: ArgumentType = defaultValue,
    argument2: AnotherArgumentType
): ReturnType {
    // body
}
```

File: src/main/kotlin/io/github/rybalkinsd/kohttp/ext/HttpContextExt.kt (hunk 3 above)
Comment (DeviantBadge) [reply src/main/kotlin/io/github/rybalkinsd/kohttp/ext/HttpContextExt.kt]: Code style changes, it was too wide and uncomfortable to read

File: src/main/kotlin/io/github/rybalkinsd/kohttp/ext/StringExt.kt (hunk 4 above)
Comment (DeviantBadge) [reply src/main/kotlin/io/github/rybalkinsd/kohttp/ext/StringExt.kt]: Same as before - https://github.com/rybalkinsd/kohttp/pull/99#discussion_r289618474

File: src/main/kotlin/io/github/rybalkinsd/kohttp/pool/CoroutineExecutorService.kt (hunk 6 above)
Comment (DeviantBadge) [reply src/main/kotlin/io/github/rybalkinsd/kohttp/pool/CoroutineExecutorService.kt]: will make it later

File: src/main/kotlin/io/github/rybalkinsd/kohttp/util/json.kt (hunk 7 above)
Comment (DeviantBadge) [reply src/main/kotlin/io/github/rybalkinsd/kohttp/util/json.kt]: Lambda formatting - https://kotlinlang.org/docs/reference/coding-conventions.html#lambda-formatting
Tabulation i will change back

File: src/test/kotlin/io/github/rybalkinsd/kohttp/Utils.kt (hunk 8 above)
Comment (DeviantBadge) [reply src/test/kotlin/io/github/rybalkinsd/kohttp/Utils.kt]: My fault, Idea was configured with such breaks by default

File: src/test/kotlin/io/github/rybalkinsd/kohttp/ext/StringAsyncExtTest.kt (hunk 10 above)
Comment (DeviantBadge) [reply src/test/kotlin/io/github/rybalkinsd/kohttp/ext/StringAsyncExtTest.kt]: Tests are better grouped up in file list, because files are placed in alphabetical order

File: src/main/kotlin/io/github/rybalkinsd/kohttp/dsl/UploadDsl.kt [hunk 14]
```diff
@@ -6,6 +6,32 @@ import okhttp3.Call
 import okhttp3.Response
 
 /**
+ * Method provides an synchronous DSL call of HTTP POST to UPLOAD file
+ *
+ * @return a `Response` instance.
+ *
+ * Usage example using the default `defaultHttpClient`:
```
[the remark this answers is not in the document]
Comment (DeviantBadge) [reply src/main/kotlin/io/github/rybalkinsd/kohttp/dsl/UploadDsl.kt]: How do you think is better to say it?
